File: library/src/main/java/com/github/library/response/ResponseFileHandler.java
package com.github.library.response;

import android.os.Handler;
import android.os.Looper;

import com.github.library.Interface.ResultHandler;
import com.github.library.enums.ErrorCode;


/**
 * Created by alishatergholi on 12/16/17.
 */
public abstract class ResponseFileHandler extends ResultHandler {

    @Override
    protected void onSuccess(String url, byte[] result) {
        new Handler(Looper.getMainLooper()).post(() -> onSuccess(result));
    }

    @Override
    public void onProgress(final double percent, final long bytesWritten, final long totalSize) {
        super.onProgress(percent, bytesWritten, totalSize);
        new Handler(Looper.getMainLooper()).post(() -> onProgress(percent, bytesWritten, totalSize));
    }

    @Override
    public void onFailure(String url, final ErrorCode errorCode) {
        new Handler(Looper.getMainLooper()).post(() -> onFailure(errorCode.getCode(), errorCode.getDescription()));
    }

    protected abstract void onSuccess(byte[] result);

    public abstract void onFailure(int errorCode, String errorMsg);


}

File: library/src/main/java/com/github/library/response/ResponseFileHandler.kt
package com.github.library.response

import android.os.Handler
import android.os.Looper

import com.github.library.helper.ResultHandler
import com.github.library.enums.ErrorCode
import okio.Buffer


/**
 * Created by alishatergholi on 12/16/17.
 */
abstract class ResponseFileHandler : ResultHandler() {

    override fun onSuccess(url: String, result: Buffer) {
        Handler(Looper.getMainLooper()).post { onSuccess(result.readByteArray()) }
    }

    override fun onProgress(percent: Double, bytesWritten: Long, totalSize: Long) {
        super.onProgress(percent, bytesWritten, totalSize)
        Handler(Looper.getMainLooper()).post { onProgress(percent, bytesWritten, totalSize) }
    }

    override fun onFailure(url: String, errorCode: ErrorCode) {
        Handler(Looper.getMainLooper()).post { onFailure(errorCode.code, errorCode.description) }
    }

    protected abstract fun onSuccess(result: ByteArray)

    protected abstract fun onFailure(errorCode: Int, errorMsg: String)


}